p2p/src/lib.rs

```diff
@@ -117,24 +184,90 @@ where
         event: ConnectivityEvent<NetworkingBackend>,
     ) -> error::Result<()> {
         match event {
-            ConnectivityEvent::Accept(socket) => self.create_peer(socket, PeerRole::Inbound),
-            ConnectivityEvent::Connect(address) => self
-                .network
-                .connect(address)
-                .await
-                .map(|socket| self.create_peer(socket, PeerRole::Outbound))?,
+            ConnectivityEvent::Accept(peer_id, socket) => {
+                self.create_peer(peer_id, socket, PeerRole::Inbound)
+            }
+            ConnectivityEvent::Connect(address) => {
+                self.network.connect(address).await.map(|(peer_id, socket)| {
+                    self.create_peer(peer_id, socket, PeerRole::Outbound)
+                })?
+            }
         }
 
         Ok(())
     }
 
-    fn peer_discovered(&mut self, peers: &[NetworkingBackend::Address]) -> error::Result<()> {
-        println!("peers discovered: {:#?}", peers);
+    /// Try to establish new outbound connections if the total number of
+    /// active connections the local node has is below threshold
+    ///
+    // TODO: move all peer management to separate file
+    async fn auto_connect(&mut self) -> error::Result<()> {
+        // we have enough active connections
+        if self.peers.len() >= MAX_ACTIVE_CONNECTIONS {
+            return Ok(());
+        }
+
+        // we don't know of any peers
+        if self.discovered.is_empty() {
+            return Err(P2pError::NoPeers);
+        }
+
+        let npeers = std::cmp::min(
+            self.discovered.len(),
+            MAX_ACTIVE_CONNECTIONS - self.peers.len(),
+        );
+
+        // TODO: improve peer selection
+        let mut peers = Vec::new();
+        let mut iter = self.discovered.iter();
+
+        for _ in 0..npeers {
+            let peer_info = iter.next().expect("Peer to exist");
+            let (ip4, ip6) = match peer_info.1 {
+                PeerAddrInfo::Raw { ip4, ip6 } => (ip4, ip6),
+            };
+            assert!(!ip4.is_empty() || !ip6.is_empty());
+
+            // TODO: let user specify their preference?
+            let addr = if ip6.is_empty() {
+                Arc::clone(ip4.iter().next().unwrap())
+            } else {
+                Arc::clone(ip6.iter().next().unwrap())
+            };
+
+            peers.push((*peer_info.0, addr));
```

Comment:
Please use `(0..npeers).into_iter().map(...)`

(I know there's a TODO, but this is simple enough to be done in this PR.)

On the other hand, this algorithm can be improved significantly. First, there has to be some pseudo-randomness in this. Otherwise, statistically speaking, a specific group of people will be bombarded with more connections than others. So what I suggest is, pick random references to `discovered`, put them in a set (to ensure there are no duplicates), then loop over them with a proper iterator. Using indexes here isn't really necessary and is basically the reason for that unnecessary `expect()`.

Reply:
This current implementation of the peer selection algorithm is horrible, I fully admit that. Right now though, and probably for the next two months, there aren't going to be external peers so having more sophisticated mechanisms for selecting peers is not needed. I have some plans on how to improve the overall connectivity of our entire blockchain network but that requires that we have external peers (outside our local network) and a way to exchange peer information with them.

But I'll remove the iteration and replace it with `map()`

Reply:
I get you... though it's not a big deal and can be easily implemented. Anyway, if you plan to keep it, please open an issue and reference this comment so that we don't forget.

Reply:
https://github.com/mintlayer/mintlayer-core/issues/98